Mathlib/Data/Quot.lean

```diff
@@ -0,0 +1,719 @@
+    (hs : ∀ a₁ a₂ b, r a₁ a₂ → t (f a₁ b) (f a₂ b)) (q₁ : Quot r) (q₂ : Quot s) : Quot t :=
+  Quot.lift₂ (fun a b => Quot.mk t <| f a b) (fun a b₁ b₂ hb => Quot.sound (hr a b₁ b₂ hb))
+    (fun a₁ a₂ b ha => Quot.sound (hs a₁ a₂ b ha)) q₁ q₂
+
+@[simp]
+theorem map₂_mk (f : α → β → γ) (hr : ∀ a b₁ b₂, s b₁ b₂ → t (f a b₁) (f a b₂))
+    (hs : ∀ a₁ a₂ b, r a₁ a₂ → t (f a₁ b) (f a₂ b)) (a : α) (b : β) :
+    Quot.map₂ f hr hs (Quot.mk r a) (Quot.mk s b) = Quot.mk t (f a b) :=
+  rfl
+
+/-- A binary version of `quot.rec_on_subsingleton`. -/
+-- porting note: removed `@[reducible]` because it caused extremely slow `simp`
+@[elab_as_elim]
+protected def recOnSubsingleton₂ {φ : Quot r → Quot s → Sort _}
+    [h : ∀ a b, Subsingleton (φ ⟦a⟧ ⟦b⟧)] (q₁ : Quot r)
+    (q₂ : Quot s) (f : ∀ a b, φ ⟦a⟧ ⟦b⟧) : φ q₁ q₂ :=
+  @Quot.recOnSubsingleton _ r (fun q => φ q q₂)
+    (fun a => Quot.ind (β := λ b => Subsingleton (φ (mk r a) b)) (h a) q₂) q₁
+    fun a => Quot.recOnSubsingleton q₂ fun b => f a b
+
+@[elab_as_elim]
+protected theorem induction_on₂ {δ : Quot r → Quot s → Prop} (q₁ : Quot r) (q₂ : Quot s)
+    (h : ∀ a b, δ (Quot.mk r a) (Quot.mk s b)) : δ q₁ q₂ :=
+  Quot.ind (β := λ a => δ a q₂) (fun a₁ => Quot.ind (fun a₂ => h a₁ a₂) q₂) q₁
+
+@[elab_as_elim]
+protected theorem induction_on₃ {δ : Quot r → Quot s → Quot t → Prop} (q₁ : Quot r)
+    (q₂ : Quot s) (q₃ : Quot t) (h : ∀ a b c, δ (Quot.mk r a) (Quot.mk s b) (Quot.mk t c)) :
+    δ q₁ q₂ q₃ :=
+  Quot.ind (β := λ a => δ a q₂ q₃) (fun a₁ => Quot.ind (β := λ b => δ _ b q₃)
+    (fun a₂ => Quot.ind (fun a₃ => h a₁ a₂ a₃) q₃) q₂) q₁
+
+instance lift.decidablePred (r : α → α → Prop) (f : α → Prop) (h : ∀ a b, r a b → f a = f b)
+    [hf : DecidablePred f] :
+    DecidablePred (Quot.lift f h) :=
+  fun q => Quot.recOnSubsingleton (motive := λ _ => Decidable _) q hf
+
+/-- Note that this provides `decidable_rel (quot.lift₂ f ha hb)` when `α = β`. -/
+instance lift₂.decidablePred (r : α → α → Prop) (s : β → β → Prop) (f : α → β → Prop)
+    (ha : ∀ a b₁ b₂, s b₁ b₂ → f a b₁ = f a b₂) (hb : ∀ a₁ a₂ b, r a₁ a₂ → f a₁ b = f a₂ b)
+    [hf : ∀ a, DecidablePred (f a)] (q₁ : Quot r) :
+    DecidablePred (Quot.lift₂ f ha hb q₁) :=
+  fun q₂ => Quot.recOnSubsingleton₂ q₁ q₂ hf
+
+instance (r : α → α → Prop) (q : Quot r) (f : α → Prop) (h : ∀ a b, r a b → f a = f b)
+    [DecidablePred f] :
+    Decidable (Quot.liftOn q f h) :=
+  Quot.lift.decidablePred _ _ _ _
+
+instance (r : α → α → Prop) (s : β → β → Prop) (q₁ : Quot r) (q₂ : Quot s) (f : α → β → Prop)
+    (ha : ∀ a b₁ b₂, s b₁ b₂ → f a b₁ = f a b₂) (hb : ∀ a₁ a₂ b, r a₁ a₂ → f a₁ b = f a₂ b)
+    [∀ a, DecidablePred (f a)] :
+    Decidable (Quot.liftOn₂ q₁ q₂ f ha hb) :=
+  Quot.lift₂.decidablePred _ _ _ _ _ _ _
+
+end Quot
+
+namespace Quotient
+
+variable [sa : Setoid α] [sb : Setoid β]
+
+variable {φ : Quotient sa → Quotient sb → Sort _}
+
+@[inherit_doc]
+notation:arg "⟦" a "⟧" => Quotient.mk _ a
```

Comment:
Did you mean to change the meaning of this notation from mathlib3? It used to use what's now called `Quotient.mk'`, which takes the setoid as an instance argument, but this notation now only works if it can get the setoid instance from the expected type.

Reply:
I recall changing some `Quotient.mk'`s to `mk`s because instance inference didn't work well (at least, back then). Perhaps now it is safe to turn it into `mk'` again.